### ISSUE TITLE
SSL error while requesting gemini://kennedy.gemi.dev/

### ISSUE THREAD

**Issue by duncan-bayne** [omar-polo/phos#10]
While requesting gemini://kennedy.gemi.dev/ with:

```
(phos/gemini:request "gemini://kennedy.gemi.dev/")
```

... I receive an SSL error:

```
A failure in the SSL library occurred on handle #.(SB-SYS:INT-SAP #X802D9A000) (SSL_get_error: 1). ERR_print_errors(): 001E600308000000:error:0A000126:SSL routines:ssl3_read_n:unexpected eof while reading:/usr/src/crypto/openssl/ssl/record/rec_layer_s3.c:308:
   [Condition of type CL+SSL::SSL-ERROR-SSL]

Backtrace:
  0: (CL+SSL::SSL-SIGNAL-ERROR #.(SB-SYS:INT-SAP #X802D9A000) #<FUNCTION CL+SSL::SSL-READ> 1 0)
  1: ((:METHOD TRIVIAL-GRAY-STREAMS:STREAM-READ-SEQUENCE (CL+SSL::SSL-STREAM T T T)) #<unavailable argument> #<unavailable argument> #<unavailable argument> #<unavailable argument>) [fast-method]
  2: ((:METHOD STREAM-READ-SEQUENCE (TRIVIAL-GRAY-STREAMS:FUNDAMENTAL-INPUT-STREAM T)) #<CL+SSL::SSL-STREAM for #<FD-STREAM for "socket 192.168.1.17:55400, peer: 64.149.155.184:1965" {11030091C3}>> #(35 32..
  3: ((FLET FLEXI-STREAMS::FILL-BUFFER :IN FLEXI-STREAMS::READ-SEQUENCE*) 1024)
  4: ((:METHOD FLEXI-STREAMS::READ-SEQUENCE* (FLEXI-STREAMS::FLEXI-UTF-8-FORMAT T T T T)) #<FLEXI-STREAMS::FLEXI-UTF-8-FORMAT (:UTF-8 :EOL-STYLE :LF) {1102DF97F3}> #<FLEXI-STREAMS:FLEXI-IO-STREAM {1102DF97..
  5: ((SB-PCL::EMF TRIVIAL-GRAY-STREAMS:STREAM-READ-SEQUENCE) #<unused argument> #<unused argument> #<FLEXI-STREAMS:FLEXI-IO-STREAM {1102DF9783}> "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0..
  6: ((:METHOD STREAM-READ-SEQUENCE (TRIVIAL-GRAY-STREAMS:FUNDAMENTAL-INPUT-STREAM T)) #<FLEXI-STREAMS:FLEXI-IO-STREAM {1102DF9783}> "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0..
  7: (PHOS/GEMINI::READ-ALL-STRING #<FLEXI-STREAMS:FLEXI-IO-STREAM {1102DF9783}>)
  8: ((:METHOD PHOS/GEMINI:REQUEST (T)) "gemini://kennedy.gemi.dev/") [fast-method]
  9: (SB-INT:SIMPLE-EVAL-IN-LEXENV (PHOS/GEMINI:REQUEST "gemini://kennedy.gemi.dev/") #<NULL-LEXENV>)
 10: (EVAL (PHOS/GEMINI:REQUEST "gemini://kennedy.gemi.dev/"))
```

This doesn't appear on all requests.  E.g. I can request `gemini://duncan.bayne.id.au/` and it's fetched without any errors.

Unsure at this point whether this is an SSL error with `kennedy.gmi.dev`, or an OpenSSL bug, or an issue with phos.

### versions

* FreeBSD 14.1-RELEASE-p4
* GNU Emacs 29.4
* SBCL 2.4.8
* OpenSSL 3.0.13 30 Jan 2024 (Library: OpenSSL 3.0.13 30 Jan 2024)

**Comment by duncan-bayne** [comment]
I _think_ this indicates an unexpected termination by the server.  I'll let the maintainer of the server know, in case this is something they can / should fix on their end.

**Comment by omar-polo** [comment]
From the error message it seems to me that the server is not using `close_notify`.  i.e. the server is not closing the tls layer correctly.  There is one fairly popular server implementation that does this (drew devault' server, don't remember the name right now) and is also no longer maintained, but can't say it's that for sure.

I'll look better into it later anyway, thank you for the report!

**Comment by duncan-bayne** [comment]
I believe there's an option in OpenSSL to hack around it - `SSL_OP_IGNORE_UNEXPECTED_EOF`.  See https://github.com/openssl/openssl/issues/18866#issuecomment-1194219601.

**Comment by duncan-bayne** [comment]
And, thank you for the library :)  I'm planning to use it to drive https://mastodon.bsd.cafe/@daily_gemini_capsule

**Comment by acidus99** [comment]
Yep, `kennedy.gemi.dev` does not properly send `close_notify`. It's built on .NET, and last time I checked that level of control of TLS wasn't exposed. Let me see if things have changed. Regardless, there are a fair amount of capsules that don't properly send these. While technically this is a violation of the spec, it is probably something a client library needs to tolerate. 

I discuss this and other spec and implementation oddities in detail here: gemini://gemi.dev/weird.gmi

**Comment by omar-polo** [comment]
@acidus99 thanks for confirming.  I knew that a fair amount of servers don't properly use close_notify, and in fact in my other two clients I'm still treating this condition as "a warning" and doesn't raise an error.  I just forgot to do the same here (and in my clojure library as well probably.)  Also, thanks for linking your oddities lists, I'll take some inspiration from it for some additional tests!

@duncan-bayne I agree that we should somehow paper over this, but I'm still unsure how.  Ideally, these servers should be eventually fixed and in a hypothetical future we should be able to remove this workaround.  I see three possible way to improve phos:

- just ignore that specific error, in N years/decades we could revisit this choice
- add a `:ignore-missing-close-notify` parameter, then flip its default in N years/decades
- add a restart-case around it: this way it's easier for consumers to choose where to see this case and eventually how to handle it.  An interactive client may want to inform the user, while a non-interactive one may opt-in to ignore the issue.

I'm propending for the last one, but I'm open to suggestions :)

Thank you!

**Comment by acidus99** [comment]
Thanks for this @duncan-bayne and @omar-polo. This was the motivation I needed to research into this and see if things had improved and [yes they have](https://github.com/dotnet/runtime/issues/17389#issuecomment-265542133).  I updated my code so that Kennedy is properly sending a `close_notify` alert. Mozz's portal shows it's now working:

<img width="537" alt="image" src="https://github.com/user-attachments/assets/c4cab8cd-9f11-4e23-8e9e-c0dfadf44e55">

@omar-polo if you do fix this, I'm happy to setup a older version that doesn't properly set `close_notify` so you can test your fix

**Comment by duncan-bayne** [comment]
Thanks @acidus99!  That works a treat, and I can now retrieve `gemini://kennedy.gemi.dev/` with phos.
